Add input validation on follow up modal

ryancarlson opened the pull request:
- add a UISnuffleButtonTapDelegate method to handle validation on a positive button
 - when true, all input fields will be validated
 - if validation passes, then listeners on that button are fired
    - followup:subscribed is not posted to notification center, instead handled w/in the class
 - email addresses are validated automatically
 - otherwise a regex can be provided

michaelharro commented on GTViewController/Classes/Custom Classes/GTInputFieldView.m:
```diff
@@ -110,4 +127,31 @@ - (NSString *)inputFieldValue {
+
+- (NSString *) validationMessage {
+    if (![[self inputFieldValue] length]) {
+        return [NSString stringWithFormat:@"%@ was empty.", self.inputFieldLabel.text];
```
These strings need to be translated

ryancarlson replied:
👍

michaelharro commented on GTViewController/Classes/Custom Classes/UISnuffleButton.m:
```diff
@@ -445,15 +454,19 @@ - (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
-                                                                      userInfo:@{UISnuffleButtonNotificationButtonTapEventKeyEventName: trimmedEventName}];
-                }
-            }];
+            if (self.forceValidation && self.tapDelegate && [self.tapDelegate respondsToSelector:@selector(didReceiveTapOnPositiveButton:)]) {
```
This should probably read more like
```
BOOL fieldsAreValid = YES;
if (self.forceValidation && self.tapDelegate && [self.tapDelegate respondsToSelector:@selector(validateFields)]) {
    fieldsAreValid = [self.tapDelegate validateFields];
}

if (fieldsAreValid) {
    [self.tapEvents enumerateObjectsUsingBlock:^(NSString *eventName, NSUInteger idx, BOOL *stop) {
        if (eventName) {
            NSString *trimmedEventName = [eventName stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
            [[NSNotificationCenter defaultCenter] postNotificationName:UISnuffleButtonNotificationButtonTapEvent object:self userInfo:@{UISnuffleButtonNotificationButtonTapEventKeyEventName: trimmedEventName}];
         }
    }];
}
```

ryancarlson replied:
There are a couple advantages I see to keeping it this way:
 - the logic for sending the followup:subscription is all handled w/in `GTFollowupViewController` and therefore reduces a notification that has to be observed and sent. *primary benefit*
 - the method name `didReceiveTapOnPositiveButton` follows the pattern of other delegate method names. *tiny, secondary benefit*

michaelharro replied:
What happens if content creators define a positive button outside of a followup modal and add tap events to it? The way I read it nothing happens, which is a bug. You are right that it simplifies things but at the cost of having a generalized solution. I think it would be better to change it so that it makes sense to content creators, should they try something different.

ryancarlson replied:
Nothing happens, but not in the way you mean it.  If the new selector implements `didReceiveTapOnPositiveButton` then the outcome of that method happens.  As it stands no other delegates implement that, so "nothing happens" meaning the else branch occurs.

michaelharro replied:
If you want to keep this pattern, which I don't recommend. You need to implement this method on GTPage so that if a positive button appears on the page or in a panel it still fires the events. You will also need to deal with the fact that the didReceiveTapOnPositiveButton is only called if forceValidation is set, which is not necessarily going to be true if it appears elsewhere in the xml.

michaelharro commented on GTViewController/Classes/Nav Classes/GTFollowupViewController.m:
```diff
@@ -296,4 +270,55 @@ - (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)butto
+- (void)didReceiveTapOnPositiveButton:(UISnuffleButton *)positiveButton {    
+    NSArray *inputValidationErrors = [self inputValidationErrors];
+    
+    if ([positiveButton forceValidation] && [inputValidationErrors count]) {
```
You already check forceValidation in the button. Ill be fixed if you make the change suggested above.

ryancarlson replied:
True, but it's good to recheck and not assume that another caller of this delegate added in the future has already checked.

michaelharro replied:
You are right that it should be rechecked using the current pattern. If the pattern is changed you would not get the button as a parameter and it would not get checked again.

michaelharro commented on GTViewController/Classes/Nav Classes/GTFollowupViewController.m:
```diff
@@ -296,4 +270,55 @@ - (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)butto
+    NSArray *inputValidationErrors = [self inputValidationErrors];
+    
+    if ([positiveButton forceValidation] && [inputValidationErrors count]) {
+        [[[UIAlertView alloc] initWithTitle:@"Please correct these fields:"
```
Strings need to be translated

ryancarlson replied:
👍

michaelharro commented on GTViewController/Classes/Nav Classes/GTFollowupViewController.m:
```diff
@@ -296,4 +270,55 @@ - (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)butto
+        return;
+    }
+    
+    [positiveButton.tapEvents enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
```
This shouldn't be called from here. Will be fixed if you take above suggestion

ryancarlson replied:
Why not?

michaelharro replied:
It makes the renderer more general and capable of more layouts if the event logic stays in the button class. See my comment above for an example.

michaelharro commented:
Looks good. Just need to pull out a few strings and make one change in how the delegate works so there is better abstraction of logic. Other than this it is great!

ryancarlson commented:
Made the string updates, but not tracking with the benefits of restructuring the code as you suggest.  The current layout feels more correct & streamlined to me.  Feel free to elaborate on your thoughts though.

ryancarlson commented:
@frett and I also discussed making validation opt-out, so enabled by default, for positive-buttons.  It can be disabled by setting `validation="[no|false]"` on the positive button (case-insensitive).

It will work for other positive-buttons not on a follow-up form b/c the array of fields to validate will simply be empty.

ryancarlson commented:
@michaelharro this is ready for review again.  I implemented all your feedback, as well fixed a couple bugs:
 - email address formatting would fail on capital letters
 - view adjusting to show input fields could be pushed off screen if the user continued to go back and forth b/w screens w/o dismissing the keyboard
 - i refactored the code that calculates the amount to move the view to clean it up

michaelharro commented on GTViewController/Classes/Nav Classes/GTFollowupViewController.m:
```diff
@@ -296,4 +303,44 @@ - (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)butto
+- (NSArray *)inputValidationErrors {
+    __block NSMutableArray *validationErrors = @[].mutableCopy;
+    
+    [self.followupModalView.inputFieldViews enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
```
This doesn't need to change but just as a reminder, you can add the variable casting in this line:
`[self.followupModalView.inputFieldViews enumerateObjectsUsingBlock:^(GTInputFieldView *inputFieldView, NSUInteger idx, BOOL * _Nonnull stop) {`

michaelharro commented:
👍 looks great! I love some of the methods you created to clean things up.
